### DISCUSSION TITLE
Spurious Modes in FDE with PML

### DISCUSSION THREAD

**Discussion by jan-david-fischbach** [flexcompute/tidy3d#917] [Q&A]
How do spurious modes arise in FDE simulations that use PML boundary conditions.

Going further: is there anything one can do to avoid them. In a less preferred case: what are the common properties of such modes, to filter them out. The modal power seems to be concentrated towards the PML regions. Are there other indications? E. g. E and H not matching up or similar?

Regards JD

**Comment by momchil-flex** [comment]
Unfortunately I don't know much either, I haven't studied those in detail. I would be curious too if you have some insights or can suggest ways to filter them out... I'll also ask Shanhui if he has any intuition about them.

**Comment by ianwilliamson** [comment]
You can filter dielectric waveguide modes by the real and imaginary parts of their eigenvalues. Modes concentrated in the PML will be lossy and will have a larger imaginary part than real part (assuming that the eigenvalue is the propagation constant or the effective index).

Personally, I would also recommend using the `which` arg of the SciPy solver (https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.eigs.html) to sort the returned modes by largest real part (`'LR'`). It looks like Tidy3D does not set this parameter, but it probably should:
https://github.com/flexcompute/tidy3d/blob/cff19d3262c0ec20df69e04afe3fa234127b563f/tidy3d/plugins/mode/solver.py#L636-L638

**Reply by jan-david-fischbach** [reply]
Hey I can quite intuitively see why the PML modes are very lossy. I am however not sure, why the condition $\Im{n_{eff}}>\Re{n_{eff}}$ should be strictly fulfilled. Could you maybe elaborate a bit?
Regards JD

**Reply by ianwilliamson** [reply]
The amount of loss experienced by a mode as it propagates is reflected in the imaginary part of its neff or propagation constant. I assume that you are interested in the guided modes of a dielectric waveguide here? If so, these modes should have a purely real neff (in the absence of any material loss), while the PML modes will have an imaginary component in their neff (possibly large relative to the real part).

You could also look at where these modes that you've solved for lie relative to the light line for your waveguide material. The PML modes will be above the light line and the guided modes will be below the light line. This is why sorting by "largest real part" is useful as it this will put the guided modes first. As a separate point, you really don't need a PML in your mode solver when you're interested in guided modes (the ones below your light line) because, by definition, these are guided and don't need to propagate out into free space and be absorbed by a PML.

**Reply by jan-david-fischbach** [reply]
Hey @ianwilliamson, thanks for the heads up. Unfortunately I am interested in the radiative modes of a bent dielectric waveguide. Additionally, as I am working with the eigenmode expansion method (see [meow](https://github.com/flaport/meow)) I typically need to consider modes below **and** above the light line, as these form an important part of the modal basis to express arbitrary fields in (and around) the waveguide. PML modes which are purely numerical artifacts should not play a role in this description, and their presence obstructs the proper modal decomposition...

If I read you right the $\Im{n_{eff}}&gt;\Re{n_{eff}}$ condition is not necessarily strictly enforced for PML modes, but rather a rule of thumb, correct?

**Reply by ianwilliamson** [reply]
Yes, it is a rule of thumb.

**Reply by momchil-flex** [reply]
> Personally, I would also recommend using the `which` arg of the SciPy solver (https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.eigs.html) to sort the returned modes by largest real part (`'LR'`). It looks like Tidy3D does not set this parameter, but it probably should:
> 
> https://github.com/flexcompute/tidy3d/blob/cff19d3262c0ec20df69e04afe3fa234127b563f/tidy3d/plugins/mode/solver.py#L636-L638

I have some recollection of playing with that but I don't remember exactly why it's not set. I feel like there was a reason at the time but maybe it can be reconsidered, especially if this helps @Jan-David-Black 's use case. Alternatively it would be very easy to expose it to the user.

**Reply by momchil-flex** [reply]
Now that I think about it note however that the eigenvalues (of the diagonal-epsilon solver, not the tensorial one) are `-(neff + 1j * keff)**2`, so the real part is `-neff**2 + keff**2`. That minus sign should best be moved into the matrix then to use `LR`. (or I am not sure if using `SR` will be equivalent)
